src/lib/UXDController.ts

```diff
@@ -28,97 +31,107 @@ export class UXDController {
+  public async mint({
+    amount,
     slippage,
     signer,
+    collateralToken,
   }: {
-    market: string;
-    ethAmount: BigNumber;
-    slippage: BigNumber;
+    amount: number;
+    slippage: number;
     signer: Signer;
+    collateralToken?: string;
   }): Promise<ContractTransaction> {
-    return this.controllerContract.connect(signer).mintWithEth(
-      market,
-      slippage,
-      {
-        value: ethAmount,
-      },
-    );
-  }
-
-  public redeemEth({
-    market,
-    uxdAmount,
+    const ethAmount = ethers.utils.parseEther(amount.toString());
+    const ethSlippage = ethers.utils.parseEther(slippage.toString());
+    if (collateralToken) {
+      return await this.mintWithERC20(
+        ethAmount,
+        ethSlippage,
+        signer,
+        collateralToken,
+      );
+    }
+    return await this.mintWithETH(ethAmount, ethSlippage, signer);
+  }
+
+  private mintWithERC20(
+    ethAmount: BigNumber,
+    ethSlippage: BigNumber,
+    signer: Signer,
+    collateral: string,
+  ): Promise<ContractTransaction> {
+    return this.controllerContract
+      .connect(signer)
+      .mint(this.market, collateral, ethAmount, ethSlippage);
+  }
+
+  private mintWithETH(
+    ethAmount: BigNumber,
+    ethSlippage: BigNumber,
+    signer: Signer
+  ): Promise<ContractTransaction> {
+    return this.controllerContract
+      .connect(signer)
+      .mintWithEth(this.market, ethSlippage, { value: ethAmount });
+  }
+
+  public async redeem({
+    amount,
     slippage,
     signer,
+    collateralToken,
   }: {
-    market: string;
-    uxdAmount: BigNumber;
-    slippage: BigNumber;
+    amount: number;
+    slippage: number;
     signer: Signer;
+    collateralToken: string;
   }): Promise<ContractTransaction> {
-    return this.controllerContract.connect(signer).redeemEth(market, uxdAmount, slippage);
+    const uxdAmount = ethers.utils.parseEther(amount.toString());
+    const slippageAmount = ethers.utils.parseEther(slippage.toString());
+    if (collateralToken) {
+      return await this.redeemERC20(uxdAmount, slippageAmount, signer, collateralToken)
+    }
+    return await this.redeemEth(uxdAmount, slippageAmount, signer);
+  }
+
+  private redeemERC20(
+    uxdAmount: BigNumber,
+    slippageAmount: BigNumber,
+    signer: Signer,
+    collateral: string,
+  ): Promise<ContractTransaction> {
+    return this.controllerContract
+      .connect(signer)
+      .redeem(this.market, collateral, uxdAmount, slippageAmount);
+  }
+  private redeemEth(
+    uxdAmount: BigNumber,
+    slippageAmount: BigNumber,
+    signer: Signer
+  ): Promise<ContractTransaction> {
+    return this.controllerContract
+      .connect(signer)
+      .redeemEth(this.market, uxdAmount, slippageAmount);
   }
 
   // ???Rework returned infos???
-  public async getCollateralInfo(): Promise<UXDControllerNamespace.CollateralInfoStructOutput[]> {
+  public getCollateralInfo(): Promise<UXDControllerNamespace.CollateralInfoStructOutput[]> {
```

Comment:
Is this useful?

Reply:
I was using this in my demo to show the amount of collateral deposited. Can remove it not used now in the playground.